Apps/Src/UpdateVelocity.c

```diff
@@ -48,44 +55,68 @@ void Task_UpdateVelocity(void *p_arg)
         uint8_t accelPedalPercent = Pedals_Read(ACCELERATOR);
         uint8_t brakePedalPercent = Pedals_Read(BRAKE);
 
+        //Edge detector to toggle the cruise enable on only the rising edge
+        prevEnable = currEnable; //store old state in prev
+        currEnable = Switches_Read(CRUZ_EN); //read in current state
+        if((prevEnable == OFF) && (currEnable == ON)){ //we went from off to on, so we should toggle it between on and off
+            cruiseEnable ^= 1; 
+        }
+        //no need for an else block, all following logic is only dependent on the cruiseEnable value
+        
+        //record speed when we're not already cruising
+        if(!cruiseEnable){
+            cruiseSpeed = MotorController_ReadVelocity();
+        }
         // Set brake lights
-        if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT){
+        if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT_BRAKE){
             Lights_Set(BrakeLight, ON);
         }
         else{
             Lights_Set(BrakeLight, OFF);
         }
-    
+        
+
         // Deadband comparison
-        if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT){
-            desiredVelocity = 0;
-            desiredMotorCurrent = (Switches_Read(REGEN_SW) && RegenEnable) ? REGEN_CURRENT : 0;
+        if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT_BRAKE){ //mech brake is pushed down
+            desiredVelocity = 0; //velocity setpoint becomes 0
+            cruiseEnable = OFF; //turn off cruise
+            desiredMotorCurrent = (Switches_Read(REGEN_SW) && RegenEnable) ? REGEN_CURRENT : 0; //either regen brake or 0 current brake
         } 
-        else if(Switches_Read(REGEN_SW) 
-                && (MotorController_ReadVelocity() > THRESHOLD_VEL) 
-                && RegenEnable
-                && accelPedalPercent < NEUTRAL_PEDALS_PERCENT){ 
+        else if( //One-pedal regen drive case
+                Switches_Read(REGEN_SW) //regen switch is pressed
+                && (MotorController_ReadVelocity() > THRESHOLD_VEL)  //we're above the threshold value
+                && (RegenEnable == ON) //regen is allowed
+                && accelPedalPercent < NEUTRAL_PEDALS_PERCENT //we're in the one-pedal drive range
+                && (cruiseEnable == OFF) //we're not in cruise mode
+            ){ 
             
             desiredVelocity = 0;
             // set regen current based on how much the accel pedal is pressed down
-            if (accelPedalPercent > UNTOUCH_PEDALS_PERCENT) {
-                desiredMotorCurrent = REGEN_CURRENT * (REGEN_RANGE - (accelPedalPercent - UNTOUCH_PEDALS_PERCENT)) / REGEN_RANGE;
+            if (accelPedalPercent > UNTOUCH_PEDALS_PERCENT_ACCEL) {
+                desiredMotorCurrent = REGEN_CURRENT * (REGEN_RANGE - (accelPedalPercent - UNTOUCH_PEDALS_PERCENT_ACCEL)) / REGEN_RANGE;
```

Comment:
Is desired behavior for one pedal regen that the more you press the acceleration pedal, the less it applies brake force (the less it slows down)? Is it possible that this goes negative?

Reply:
Wait never mind I figured it out, checking if accelPedalPercent > untouch pedal percent will make sure it doesn't go negative. I still would like to know desired behavior though.

Reply:
Desired behavior is that under the neutral point the less you press the pedal, the harder it brakes. The more you press, the less it breaks. So brake force increases as you lift your foot off the accelerator.